subprojects/base-services/src/main/java/org/gradle/internal/reflect/ReflectionCache.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.internal.reflect;
+
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
+public abstract class ReflectionCache<T> {
+    private final Object lock = new Object();
+    private final WeakHashMap<Class<?>, WeakReference<T>> cache = new WeakHashMap<Class<?>, WeakReference<T>>();
```

Comment:
I removed the wrapping into a `WeakReference`, which was indeed not necessary, since the value itself should be responsible for wrapping its members into weak references.